Add jacobian for hydro coupling source term

pajkosmi opened the pull request:
## Proposed changes

Calculate Jacobian for M1 source terms.  Also update M1 system to use Imex timestepper.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

pajkosmi committed:
Add static return applyable

pajkosmi committed:
Add jacobian for implicit source terms

pajkosmi committed:
Update M1Grey system

wthrowe commented:
Please rebase now that #6076 is merged.

pajkosmi commented:
> Please rebase now that #6076 is merged.

rebased

wthrowe commented on src/Utilities/Protocols/StaticReturnApplyable.hpp:
```diff
@@ -0,0 +1,46 @@
+    static_assert(
+        tmpl::wrap<
+            tmpl::append<
+                tmpl::list<decltype(ConformingType::apply)>,
```
Test this with a struct where the `apply` function is a template.  I think the current form will not work, but it can be fixed by making a version of `CREATE_IS_CALLABLE` that requires the function to be static.  (By using `T::METHOD_NAME` instead of `std::declval<T>().METHOD_NAME` in the implementation.)

pajkosmi replied:
edit: After having a conversation with @nilsvu, he mentioned this shouldn't work when BaseTags are present, since they do not know their type.  I am going to proceed by keeping the first three `static_asserts` that check for argument_tags, return_tags, and no duplicate return_tags. That should work independently of templated or nontemplated apply functions.  What are your thoughts?

wthrowe replied:
Good point.  That's sounds good.

wthrowe commented on src/Utilities/Protocols/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+# Distributed under the MIT License.
```
Include this file from the parent `CMakeLists.txt`.

wthrowe commented on tests/Unit/Utilities/Protocols/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+# Distributed under the MIT License.
```
Include this file from the parent `CMakeLists.txt`.

wthrowe commented on src/Evolution/Executables/RadiationTransport/M1Grey/CMakeLists.txt:
```diff
@@ -24,6 +24,7 @@ target_link_libraries(
   GeneralRelativitySolutions
   Hydro
   Informer
+  Imex
```
Alphabetize.

wthrowe commented on tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_M1HydroCoupling.cpp:
```diff
@@ -23,8 +23,20 @@ SPECTRE_TEST_CASE("Unit.RadiationTransport.M1Grey.M1HydroCoupling",
       typename RadiationTransport::M1Grey::ComputeM1HydroCoupling<
```
Add the sector in this commit and call `TestHelpers::imex::test_sector`.

pajkosmi replied:
Would something like `Test_System.cpp` on the below branch suffice?
https://github.com/pajkosmi/spectre/blob/grey_m1/tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp

wthrowe replied:
Yes, something like that.

pajkosmi replied:
Test_System.cpp has been added to the branch.

wthrowe commented on tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/ComputeM1HydroCouplingJacobian.py:
```diff
@@ -0,0 +1,391 @@
+# Distributed under the MIT License.
```
I'm not convinced this python test is worth much.  It's obviously a copy-paste of the c++ source, so any math errors are going to be the same in both.  The IMEX `test_sector` will test this better.

pajkosmi replied:
I think the fuzzy testing has a small value to check for edge cases producing 0s in denominators.  Nevertheless, I agree with the similarity between the .py and .cpp code, so the test_sector (mentioned below) will be included.

wthrowe commented on src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp:
```diff
@@ -0,0 +1,54 @@
+/// tensors in the implicit sector.
+///
+/// \dontinclude Test_SolveImplicitSector.cpp
+/// \skip struct source_jacobian
```
This isn't showing up in the rendered version.  (There is no `struct source_jacobian` in that file.)

pajkosmi replied:
I changed it from `source_jacobian` to `jacobian`.

wthrowe commented on src/Evolution/Systems/RadiationTransport/M1Grey/M1HydroCoupling.hpp:
```diff
@@ -49,8 +54,28 @@ void compute_m1_hydro_coupling_impl(
+    const tnsr::II<DataVector, 3>& inverse_spatial_metric);
 }  // namespace detail
 
+// FIXME remove list?
```
Make a decision and do it.

wthrowe commented on src/Evolution/Systems/RadiationTransport/M1Grey/M1HydroCoupling.hpp:
```diff
@@ -78,10 +103,15 @@ struct ComputeM1HydroCoupling;
+      tt::ConformsTo<protocols::StaticReturnApplyable> {
+  // FIXME remove tags?
+  //  Tags::M1HydroCouplingNormal<NeutrinoSpecies>...,
+  //  Tags::M1HydroCouplingSpatial<Frame::Inertial, NeutrinoSpecies>...>;
```
Remove comment.

wthrowe commented on src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp:
```diff
@@ -192,22 +200,27 @@ struct EvolutionMetavars {
               evolution::dg::Actions::ApplyBoundaryCorrectionsToTimeDerivative<
                   system, volume_dim, false>,
               Actions::RecordTimeStepperData<system>,
               evolution::Actions::RunEventsAndDenseTriggers<tmpl::list<>>,
```
Remove this action.  You added a copy with the correct arguments two lines below.

wthrowe commented on src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp:
```diff
@@ -192,22 +200,27 @@ struct EvolutionMetavars {
+                  tmpl::list<imex::ImplicitDenseOutput<system>>>,
               Actions::UpdateU<system>>>,
       Actions::CleanHistory<system, local_time_stepping>,
+      imex::Actions::DoImplicitStep<system>,  // FIXME at end?
```
One line earlier, I think, actually.

wthrowe commented on src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp:
```diff
@@ -192,22 +200,27 @@ struct EvolutionMetavars {
+                               ComputeM1Closure<neutrino_species>>/*,
       Actions::MutateApply<typename RadiationTransport::M1Grey::
-                               ComputeM1HydroCoupling<neutrino_species>>>>;
+                               ComputeM1HydroCoupling<neutrino_species>>*/>>;
```
Remove.

wthrowe commented on src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp:
```diff
@@ -223,10 +236,15 @@ struct EvolutionMetavars {
+      //   Actions::MutateApply<
+      //       typename
+      //       RadiationTransport::M1Grey::ComputeM1HydroCouplingJacobian<
+      //           neutrinos::ElectronNeutrinos<1>>>,
```
Remove.

wthrowe commented on src/Evolution/Systems/RadiationTransport/M1Grey/System.hpp:
```diff
@@ -84,6 +90,58 @@ struct System<tmpl::list<NeutrinoSpecies...>> {
+  struct ImplicitSector : tt::ConformsTo<imex::protocols::ImplicitSector> {
+    using tensors = tmpl::list<Tags::TildeE<Frame::Inertial, Species>,
+                               Tags::TildeS<Frame::Inertial, Species>>;
+    // using helper = ComputeM1Closure<tmpl::list<Species>>;
```
Remove all the outdated commented stuff in this file.

wthrowe commented on src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp:
```diff
@@ -9,6 +9,8 @@
 #include <utility>  // IWYU pragma: keep  // for move
```
What is the reason for the changes in this commit?  We don't want to do this if we can avoid it.

pajkosmi replied:
In `M1HydroCoupling.hpp` the `return_tags` for the `ComputeM1HydroCoupling` struct and `ComputeM1HydroCouplingJacobian` struct need these tags.  Without these changes, the compiler error says the return tags cannot be found in the databox.

wthrowe replied:
Those are outputs, though.  Why do you need to compute them?

pajkosmi replied:
We will compute them in the EvolveSystem (below link).  My understanding is, even if they're return_tags, they need to be in the Databox?  In `Initialize.hpp`, I see 
```
using m1_variables_tag = typename System::primitive_variables_tag;
```
Do the tags need to be prepended with `::Source` and `::jacobian` in System.hpp and then added to `primitive_variables_tag`?

https://github.com/sxs-collaboration/spectre/blob/887e81c4a148d04634b913185a2ecd368d142152/src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp#L242

wthrowe replied:
But *why* are you computing them?  What uses them?

Let me make an analogy with runtime code:
```c++
const auto energy = ...;
Scalar<DataVector> energy_source{};
compute_source(make_not_null(&energy_source), energy);
return energy;
```
I am asking the equivalent of "Why is `energy_source` needed?" and getting the answer "It's a required argument of the the `compute_source` function.".  As far as I can tell, you write to these variables, but never read from them, which doesn't seem useful.

pajkosmi replied:
Following the SpECTRE discussion, I understand your question now.  The short answer is we do not need these changes, yet.  Formally, this PR does not couple the neutrinos to the fluid.  So nothing, in this PR, needs those terms.  In the future, they will be needed.   

The reason why I included these changes is I want to see if the changes on the original M1 branch were able to compile.  Specifically, in the EvolveM1Grey file, the initialization and step action calls to HydroCoupling were commented out. 

https://github.com/wthrowe/spectre/blob/b73281b3c6c27e029d55097d4fec1f6144bd04a3/src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp#L433

edit (italics is not needed): _After including them in the Evolve system, I found it could not compile, which led me to discover the tags are needed in the databox.  For this PR, I will not include these changes.  For posterity, (myself mostly) the tags for Jacobians of the stiff hydro coupling sources will need to be included._

wthrowe replied:
They were commented out on the branch because they were left over from the old non-IMEX version.  They were supposed to be deleted when I cleaned up the branch, not uncommented.

You do not need the sources and jacobians in the DataBox in order to couple to fluids.  You can calculate them during the computation of the hydro sources without permanently storing them.

pajkosmi commented:
Ping, I freshly rebased and addressed the reviewer comments.

wthrowe commented:
Don't delete the `CMakeLists.txt` files, or at least the one in `src`.  That's preventing some of the CI checks from running.

wthrowe commented:
The `CMakeLists.txt` files you had before the fixup were fine, you were just missing the calls to `add_subdirectory` in the parents.

wthrowe commented on tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp:
```diff
@@ -0,0 +1,117 @@
+
+SPECTRE_TEST_CASE("Unit.RadiationTransport.M1Grey.System.Imex",
+                  "[Unit][M1Grey]") {
+  // struct DummySpecies;
```
Clean up things like this.

wthrowe commented on src/Utilities/CMakeLists.txt:
```diff
@@ -115,6 +115,7 @@ endif()
 
 add_subdirectory(ErrorHandling)
 add_subdirectory(StdHelpers)
+add_subdirectory(Protocols)
```
Alphabetize.

wthrowe commented on tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp:
```diff
@@ -0,0 +1,117 @@
+  tenex::evaluate(make_not_null(&tilde_e),
+                  sqrt(tilde_s(ti::i) * tilde_s(ti::j) *
+                       inverse_spatial_metric(ti::I, ti::J)) +
+                      0.01);  // force nearly thin
```
Don't just test the thin limit.

wthrowe commented on tests/Unit/Utilities/Protocols/Test_StaticReturnApplyable.cpp:
```diff
@@ -53,4 +69,7 @@ SPECTRE_TEST_CASE("Unit.Utilities.Protocols.StaticReturnApplyable",
       T1{}, T2{}, T3{}, T4{});
   db::mutate_apply<Applyable>(make_not_null(&box));
   CHECK(true);
```
Only need one `CHECK(true)`.  It's just there to avoid a catch error when there are no checks.

wthrowe commented on src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp:
```diff
@@ -29,7 +29,7 @@ namespace imex::protocols {
 ///
 /// \dontinclude Test_SolveImplicitSector.cpp
-/// \skip struct source_jacobian
+/// \skip struct jacobian
```
This example is showing up as
```c++
    struct jacobian {
      using return_tags = tmpl::list<>;
      using argument_tags = tmpl::list<>;
 
      static void apply() {}
    };
```
which doesn't conform to the protocol and isn't a very useful example in any case.

Actually, it looks like nothing actually uses the new protocols, so that should be fixed too.

pajkosmi replied:
This was partly a case-sensitive issue.  `struct Jacobian` (further above) has an implementation that is not a stub.  I also made sure both `Jacobian` and `source` structs now inherit from `ImplicitSource` and `ImplicitSourceJacobian`.

wthrowe reviewed (changes requested):
You can squash the existing fixups.  I'll take another look at the whole thing after.

pajkosmi commented:
squashed

wthrowe commented:
You don't seem to have squashed.

pajkosmi commented:
> You don't seem to have squashed.

edit: see below and ignore italics: _When you say squash, do you mean squash all the fixups into 1 fixup commit (like it is now) or squash the 6 current commits (5 initial commits + 1 fixup commit) into 1 commit?_

pajkosmi commented:
Regarding my previous comment, now I (think) I understand what you meant by squash.  I have squashed the fixup commits into the previous relevant commits and rebased.

wthrowe commented on src/Evolution/Imex/Protocols/ImplicitSector.hpp:
```diff
@@ -93,7 +93,7 @@ namespace imex::protocols {
 ///
 /// \snippet Test_SolveImplicitSector.cpp source
-/// \snippet Test_SolveImplicitSector.cpp jacobian
+/// \snippet Test_SolveImplicitSector.cpp Jacobian
```
Mention/link to the new protocols from the docs in this file, and check in this protocol's `test` that the sub-structs conform to the new protocols.

wthrowe commented on src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp:
```diff
@@ -0,0 +1,54 @@
+///
+/// \dontinclude Test_SolveImplicitSector.cpp
+/// \skip struct Jacobian
+/// \until };
```
Use the snippet that is used on the `SolveImplicitSector` protocol.

pajkosmi replied:
I do not see a snippet in `SolveImplicitSector.hpp`.  Do you mean change to 
`/// \snippet Test_SolveImplicitSector.cpp Jacobian`
 (like what's used in `ImplicitSector.hpp`), or a snippet somewhere else?

pajkosmi replied:
Update: after talking at the spectre technical call, I went with the above.

wthrowe replied:
I meant just `ImplicitSector`, sorry.  You figured it out.

wthrowe commented on src/Evolution/Systems/RadiationTransport/M1Grey/M1HydroCoupling.hpp:
```diff
@@ -10,10 +10,15 @@
+#include "Evolution/Imex/Protocols/ImplicitSource.hpp"
+#include "Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp"
+#include "Evolution/Imex/Tags/Jacobian.hpp"
+#include "Evolution/Systems/RadiationTransport/M1Grey/Tags.hpp"  // IWYU pragma: keep
```
Don't add IWYU pragmas.

wthrowe commented on tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp:
```diff
@@ -0,0 +1,114 @@
+  TestHelpers::imex::test_sector<sector>(stencil_size, tolerance,
+                                         explicit_values, list_of_values);
+
+  // optically thin regime
```
Check an intermediate case, not just perfectly thick and thin.

wthrowe commented on tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp:
```diff
@@ -0,0 +1,114 @@
+  tnsr::I<DataVector, 3> fluid_velocity{};
+  get<0>(fluid_velocity) = DataVector{0.};
+  get<1>(fluid_velocity) = DataVector{0.};
+  get<2>(fluid_velocity) = DataVector{0.};
```
Don't set this to 0.

wthrowe commented on tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp:
```diff
@@ -0,0 +1,114 @@
+  get<0, 2>(spatial_metric) = DataVector{0.0};
+  get<1, 1>(spatial_metric) = DataVector{1.0};
+  get<1, 2>(spatial_metric) = DataVector{0.0};
+  get<2, 2>(spatial_metric) = DataVector{1.0};
```
Set these to something nontrivial.

wthrowe commented on src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp:
```diff
@@ -197,15 +205,16 @@ struct EvolutionMetavars {
+              evolution::Actions::RunEventsAndDenseTriggers<
+                  tmpl::list<imex::ImplicitDenseOutput<system>>>,
               Actions::UpdateU<system>>>,
+      imex::Actions::DoImplicitStep<system>,  // FIXME at end?
```
Remove comment.

wthrowe commented on tests/InputFiles/RadiationTransport/M1Grey/ConstantM1.yaml:
```diff
@@ -3,6 +3,7 @@
 
```
Presumably the previous commit doesn't work without these changes, so squash these two commits together.

wthrowe commented on src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp:
```diff
@@ -9,6 +9,8 @@
 #include <utility>
```
Drop this commit.

wthrowe reviewed (changes requested):
> Regarding my previous comment, now I (think) I understand what you meant by squash. I have squashed the fixup commits into the previous relevant commits and rebased.

Yes, sorry.  You figured it out.

pajkosmi commented:
Everything has been addressed/rebased, except the `ImplicitSourceJacobian.hpp` point I need clarification on above.

wthrowe commented on src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp:
```diff
@@ -28,9 +28,8 @@ namespace imex::protocols {
-/// \skip struct Jacobian
-/// \until };
+/// \snippet Test_SolveImplicitSector.cpp Jacobian
+
```
Remove blank line.

wthrowe commented on tests/Unit/Evolution/Imex/Test_SolveImplicitSector.cpp:
```diff
@@ -183,11 +183,15 @@ struct InitialGuess {
 
+  static_assert(
+      tt::assert_conforms_to_v<Source, imex::protocols::ImplicitSource>);
+
```
Move these (two more below) outside the structs.

wthrowe commented on src/Evolution/Imex/Protocols/ImplicitSource.hpp:
```diff
@@ -0,0 +1,44 @@
+///
+/// \dontinclude Test_SolveImplicitSector.cpp
+/// \skip struct source
+/// \until };
```
Switch this example to the preexisting one like you did for the Jacobian.

wthrowe commented on src/Evolution/Imex/Protocols/ImplicitSector.hpp:
```diff
@@ -93,7 +94,7 @@ namespace imex::protocols {
+/// \snippet Test_SolveImplicitSector.cpp Jacobian
 struct ImplicitSector {
   template <typename ConformingType>
   struct test {
```
The checks that the members conform to the new protocols still need to be added in here.

wthrowe reviewed (changes requested):
The test of the system looks good now.

pajkosmi commented:
Comments addressed

wthrowe commented:
Looks good.  The test timeouts look unrelated to these changes, so don't worry about them.  Go ahead and squash in the remaining fixups.

pajkosmi commented:
squashed

wthrowe reviewed (approved):
Test failures look unrelated.
